src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -770,6 +770,77 @@ internal override bool IsExplicitDefinitionOfNoPiaLocalType
             }
         }
 
+        private void CheckPresenceOfTypeIdentifierAttribute()
+        {
+            // We want this function to be as cheap as possible, it is called for every top level type
+            // and we don't want to bind attributes attached to the declaration unless there is a chance
+            // that one of them is TypeIdentifier attribute.
+            ImmutableArray<SyntaxList<AttributeListSyntax>> attributeLists = GetAttributeDeclarations();
+
+            foreach (SyntaxList<AttributeListSyntax> list in attributeLists)
+            {
+                var syntaxTree = list.Node.SyntaxTree;
+                Binder binder = null;
+
+                foreach (AttributeListSyntax attrList in list)
+                {
+                    foreach (AttributeSyntax attr in attrList.Attributes)
+                    {
+                        SimpleNameSyntax simpleName = attr.Name.GetUnqualifiedName();
+                        string name = simpleName.Identifier.ValueText;
+
+                        if (name == "TypeIdentifier" || name == AttributeDescription.TypeIdentifierAttribute.Name)
+                        {
+                            // This attribute syntax might be an application of TypeIdentifierAttribute.
+                            // Let's bind it.
+                            // For simplicity we bind all attributes.
+                            GetAttributes();
+                            return;
+                        }
+
+                        if (attr.Name == simpleName)
+                        {
+                            // Check if the name matches an alias bound to the TypeIdentifierAttribute
```

Comment:
:bulb: Consider an alternate approach where you look only at the namespaces in the syntax tree. If you see a using alias directive that maps to `TypeIdentifierAttribute`, assume that the directive is used and call `GetAttributes`. You will avoid needing to check every attribute usage against aliases, further improving fast paths.

Reply:
> further improving fast paths

This is a speculation. Figuring out if alias binds to a type requires binding to occur. Checking if name matches an alias is a simple lookup in a hash table.

Reply:
I believe it is not common for an attribute application to use an alias, therefore we can completely avoid binding anything.

Reply:
For the case I mentioned, you could check to see if any alias maps to the string name "TypeIdentifierAttribute". Binding would not be necessary.

Reply:
If we find out that with this approach SourceModuleSymbol.ContainsExplicitDefinitionOfNoPiaLocalTypes still has noticeable negative effect for the customer, I'll see how can we adopt VB approach, which is completely syntax based.